hoomd/md/pytest/test_custom_force.py

```diff
@@ -0,0 +1,323 @@
+import pytest
+import numpy as np
+import numpy.testing as npt
+
+try:
+    import cupy
+    CUPY_IMPORTED = True
+except ImportError:
+    CUPY_IMPORTED = False
+
+import hoomd
+from hoomd import md
+
+
+class MyForceCPU(md.force.Custom):
+
+    def __init__(self):
+        super().__init__()
+
+    def set_forces(self, timestep):
+        with self.cpu_local_force_arrays as arrays:
+            arrays.force = -5
+            arrays.potential_energy = 37
+            arrays.torque = 23
+            arrays.virial = np.arange(6)[None, :]
+
+
+class MyForceGPU(md.force.Custom):
+
+    def __init__(self):
+        super().__init__()
+
+    def set_forces(self, timestep):
+        with self.gpu_local_force_arrays as arrays:
+            arrays.force = -5
+            arrays.potential_energy = 37
+            arrays.torque = 23
+            # cupy won't let me use two indices on arrays of length 0
+            if len(arrays.virial) > 0:
+                for i in range(6):
+                    arrays.virial[:, i] = i
+
+
+def _skip_if_cupy_not_imported_and_gpu_device(sim):
+    """Force classes cant set values if cupy isn't installed."""
+    gpu_device = isinstance(sim.device, hoomd.device.GPU)
+    if gpu_device and not CUPY_IMPORTED:
+        pytest.skip("Cannot run this test on GPU without cupy")
+
+
+def _using_gpu_with_cpu_device(force_cls, sim):
+    """Sims should throw an error if this returns true."""
+    gpu_class = force_cls.__name__.endswith("GPU")
+    cpu_device = isinstance(sim.device, hoomd.device.CPU)
+    return cpu_device and gpu_class
+
+
+def _try_running_sim(sim, tsteps):
+    """Run the sim while checking that the error is raised."""
+    should_error = _using_gpu_with_cpu_device(
+        sim.operations.integrator.forces[0].__class__, sim)
+    if should_error:
+        with pytest.raises(RuntimeError):
+            sim.run(tsteps)
+    else:
+        sim.run(tsteps)
+    return should_error
```

Comment:
If I follow this logic, this is essentially allowing us to later filter out errors which just come from accessing GPU classes on the CPU. It's mainly just for the convenience to write the test once and properly ignore these errors?

Reply:
This helper function makes sure the simulation fails to run if we are trying to use the GPU custom force class with a CPU simulation device, and otherwise runs the simulation as it is supposed to be run. It is a check that has to be done every time we run a test, so putting all the logic in one helper function is cleaner than writing the same code over and over.

Others can comment if they think there is a cleaner way to write this, I know its somewhat awkward because we have to check the return value after every time we call `_try_running_sim` so the test can exit properly.

Reply:
I think it might benefit from a more clear name. Especially because `try` is a python keyword. Perhaps something like `run_should_fail` because that's essentially what it returns. But this function is also actually doing the run as well. Will defer to others opinion.

Reply:
I'll wait until @joaander has a pass over this PR before deciding how to move forward with this comment.

Reply:
I agree, this is confusing. It is unclear what code path any given test run is supposed to take and what it is testing.

1) If you are attempting to test that running on the GPU with a CPU device raises an error, write one specific test that does just that.

2) If this is a construction to avoid invalid parameters in the matrix resulting from invalid combinations of two separate parameters, then use a different construction. For example, instead of parameterizing the test on the force class, select the force class based on the device. If that isn't possible, use skips - but skips should be used sparingly. I prefer when they are used only when the user's hardware or software configuration prevents a test from running.